node.go

```diff
@@ -18,27 +18,51 @@ import (
 	"github.com/cosmos/iavl/internal/encoding"
 )
 
+const (
+	// ModeLegacyLeftNode is the mode for legacy left child in the node encoding/decoding.
+	ModeLegacyLeftNode = 0x01
+	// ModeLegacyRightNode is the mode for legacy right child in the node encoding/decoding.
+	ModeLegacyRightNode = 0x02
+)
+
 // NodeKey represents a key of node in the DB.
 type NodeKey struct {
 	version int64
 	nonce   int32
 }
 
+// GetKey returns a byte slice of the NodeKey.
 func (nk *NodeKey) GetKey() []byte {
 	b := make([]byte, 12)
 	binary.BigEndian.PutUint64(b, uint64(nk.version))
 	binary.BigEndian.PutUint32(b[8:], uint32(nk.nonce))
 	return b
 }
 
+// GetNodeKey returns a NodeKey from a byte slice.
+func GetNodeKey(key []byte) *NodeKey {
+	return &NodeKey{
+		version: int64(binary.BigEndian.Uint64(key)),
+		nonce:   int32(binary.BigEndian.Uint32(key[8:])),
+	}
+}
+
+// GetRootKey returns a byte slice of the root node key for the given version.
+func GetRootKey(version int64) []byte {
+	b := make([]byte, 12)
+	binary.BigEndian.PutUint64(b, uint64(version))
+	binary.BigEndian.PutUint32(b[8:], 1)
+	return b
+}
+
 // Node represents a node in a Tree.
 type Node struct {
 	key           []byte
 	value         []byte
 	hash          []byte
 	nodeKey       *NodeKey
-	leftNodeKey   *NodeKey
-	rightNodeKey  *NodeKey
+	leftNodeKey   []byte
+	rightNodeKey  []byte
```

Comment:
🤩

Reply:
I propose taking this further.  Why not defined `nodeKey as []byte` and store version and nonce as private, top-level fields on `Node`?

Reply:
`[]byte nodeKey` is meaningless, `leftNodeKey` and `rightNodeKey` is used in getting children
The only thing we need is `nonce` and `version`

Reply:
`NodeKey` seems also without meaning, just makes the code harder to read.  if we need `nonce` and `version` why not put it on `Node` directly?

Reply:
good question, the nonce and version are used together in several encodings, so I created the specific struct `NodeKey`
and also it ensures `version+nonce` as a node key